segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/AsyncBaseChunkStorage.java

```diff
@@ -0,0 +1,637 @@
+ * This is indicated by {@link ChunkStorage#supportsAppend()}. For example S3 compatible Chunk Storage this would return false. </li>
+ * <li> Does {@link ChunkStorage}  support for concatenating chunks? This is indicated by {@link ChunkStorage#supportsConcat()}.
+ * If this is true then concat operation concat will be invoked otherwise append functionality is invoked.</li>
+ * <li>In addition {@link ChunkStorage} may provide ability to truncate chunks at given offsets (either at front end or at tail end). This is indicated by {@link ChunkStorage#supportsTruncation()}. </li>
+ * </ul>
+ * There are some obvious constraints - If ChunkStorage supports concat but not natively then it must support append .
+ *
+ * For concats, {@link ChunkStorage} supports both native and append, ChunkedSegmentStorage will invoke appropriate method depending on size of target and source chunks. (Eg. ECS)
+ *
+ * The implementations in this repository are tested using following test suites.
+ * <ul>
+ * <li>SimpleStorageTests</li>
+ * <li>ChunkedRollingStorageTests</li>
+ * <li>ChunkStorageProviderTests</li>
+ * <li>SystemJournalTests</li>
+ * </ul>
+ */
+@Slf4j
+@Beta
+public abstract class AsyncBaseChunkStorage implements ChunkStorage {
+
+    private final AtomicBoolean closed;
+
+    private final Executor executor;
+
+    /**
+     * Constructor.
+     *
+     * @param executor  An Executor for async operations.
+     */
+    public AsyncBaseChunkStorage(Executor executor) {
+        this.closed = new AtomicBoolean(false);
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+    }
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports truncate operation on chunks.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsTruncation();
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports append operation on chunks.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsAppend();
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports merge operation either natively or through appends.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsConcat();
+
+    /**
+     * Determines whether named file/object exists in underlying storage.
+     *
+     * @param chunkName Name of the chunk to check.
+     * @return True if the object exists, false otherwise.
+     */
+    @Override
+    final public CompletableFuture<Boolean> exists(String chunkName) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        // Validate parameters
+        checkChunkName(chunkName);
+
+        long traceId = LoggerHelpers.traceEnter(log, "exists", chunkName);
+        // Call concrete implementation.
+        val returnFuture =  checkExistsAsync(chunkName);
+        returnFuture.thenApplyAsync(retValue -> {
```

Comment:
In case of exception you aren't going to execute this callback.
You should use whenComplete (not whenCompleteAsync)

and return the resulting CompletableFuture

Reply:
This is simply following the existing pattern of not logging trace leave or metrics in case of failures.

Reply:
Makes sense.

Reply:
done.